Candidate for version 1

lohathe opened the pull request:
Merge of several issues:
(o) improve performance and leverage worker threads
(o) improve usability and UX
(o) several minor fixes

lohathe committed:
Refactor model for perfomance issues.

* add internal status to nodes to avoid null updates
* use scandir instead of fwalk

lohathe committed:
Add performance test to quantity the gain of scandir wrt fwalk.

lohathe committed:
Fix signal parameter.

lohathe committed:
Add simple output for excluded paths/folders.

lohathe committed:
Add basic action to main window.

lohathe committed:
Make toolbar button toggleable. Update status messages for actions.

lohathe committed:
Add useful information for the user.

lohathe committed:
Show number of included nodes (files and dirs) vs total nodes.

lohathe committed:
Clear filters list when changing root folder.

lohathe committed:
Fix minor bugs.

* exit the save dialog does not generate an exception.
* clear the list of matched folders/files when changing root.

lohathe committed:
Add feature to match/not match the root path.

lohathe committed:
Minor fix. Corner case to update view.

lohathe committed:
Merge branch 'issue-6-large-folders-performance'

lohathe committed:
Merge branch 'issue-7-user-input-output'

lohathe committed:
Make model independent from Qt

* SystemTreeNode is no more a QObject
* signals replaced with function handlers
* updated tests

lohathe committed:
Fix memory leak.

lohathe committed:
Add worker thread to make GUI responsive.

naufraghi commented on backup-excluder.py:
```diff
@@ -2,11 +2,24 @@
+def hack_for_reactive_gui():
+    global count
+    count += 1
+    if not (count & 0x3F):
```
Have you considered [itertools.cycle](https://docs.python.org/2/library/itertools.html#itertools.cycle)? But more deeply, are you sure that this call is faster than a plain `QCoreApplication.processEvents()`?

naufraghi commented on backup-excluder.py:
```diff
@@ -36,48 +48,102 @@ def _colorBackground(self, brush):
         return root
 
 
+class WorkerThread(threading.Thread):
```
Things can become complicated mixing Python `Thread`s and  `QThread`s, do you need both?

naufraghi commented on backup-excluder.py:
```diff
@@ -88,29 +154,174 @@ def customInit(self, basePath):
+        self.tree.clear()
+        self.output.clear()
+
+    def _setOutputEnabled(self, enabled):
```
Have you considered disabling the parent widget? [QWidget.setEnabled](http://doc.qt.io/qt-5/qwidget.html#enabled-prop):

> Disabling a widget implicitly disables all its children. Enabling respectively enables all child widgets unless they have been explicitly disabled. It it not possible to explicitly enable a child widget which is not a window while its parent widget remains disabled.

naufraghi commented on backup-excluder.py:
```diff
@@ -88,29 +154,174 @@ def customInit(self, basePath):
         filterExpr = '(' + '|'.join(filters) + ')'
         if filterExpr == "()":
             filterExpr = "^$"
         cutFunction = re.compile(filterExpr).match
```
Have you considered the differences between [`re.match` and `re.search`](https://docs.python.org/3/library/re.html?highlight=re%20match#search-vs-match)?

naufraghi commented on backup-excluder.py:
```diff
@@ -88,29 +154,174 @@ def customInit(self, basePath):
         filters = filter(str.strip, self.edit.document().toPlainText().split("\n"))
         filterExpr = '(' + '|'.join(filters) + ')'
         if filterExpr == "()":
             filterExpr = "^$"
```
Do you really need a strange regex to use `match` inside `update` or you can also go with a `lambda x: False` if there are no filters?

naufraghi commented on model.py:
```diff
@@ -42,45 +41,117 @@ def addChild(self, child):
         self.size += child.size
-        child.parent = self
+        child.parent = weakref.ref(self)
         sup = self.parent
```
Perhaps moving the call from `sup()` to `parent()` simplifies a bit the readability:

``` python
sup = self.parent()
while isinstance(sup, SystemTreeNode):
    sup.size += child.size
￼    sup = sup.parent()
```

naufraghi commented on backup-excluder.py:
```diff
@@ -88,29 +154,174 @@ def customInit(self, basePath):
+
+    def _createSystemTree(self, initialPath):
+        self._createSystemTreeAsyncStart(initialPath)
+        return
```
leftover

naufraghi commented on backup-excluder.py:
```diff
@@ -88,29 +154,174 @@ def customInit(self, basePath):
+    def _createSystemTreeAsyncStart(self, initialPath):
+        self._notifyStatus("Please wait...reading file system. It may take a while.")
+        self._setOutputEnabled(False)
+        del self.root
```
Why are you deleting the `root` manually? This can cause a latent `AttributeError` accessing the attribute.

naufraghi commented on backup-excluder.py:
```diff
@@ -88,29 +154,174 @@ def customInit(self, basePath):
+        newRootFolder = ""
+        if fileDialog.exec():
+            newRootFolder = fileDialog.selectedFiles()
+            if len(newRootFolder) != 1:
```
Qt usually offers shortcuts for the most common usages: [QFileDialog.getExistingDirectory](http://doc.qt.io/qt-5/qfiledialog.html#getExistingDirectory)

naufraghi commented on model.py:
```diff
@@ -42,45 +41,117 @@ def addChild(self, child):
+        return (totalSize, totalNodes)
+
+    @staticmethod
+    def _createSystemTreeRecursive2(rootPath):
```
I think you can cleanup the unwanted recursion implementation.

naufraghi commented on model.py:
```diff
@@ -42,45 +41,117 @@ def addChild(self, child):
+            self.visibilityChangedHandler(newStatus, newSize)
+
+    def _set_exclusion_state_recursive(self, exclusionState):
+        self.lastState = exclusionState
```
Doesn't `lastState` seems more to be the current state? And what kind of state? Let the name be as explicit as you can.

naufraghi commented on model.py:
```diff
@@ -42,45 +41,117 @@ def addChild(self, child):
+                # the original size of the node with the size of the subtree
+                # is the little hack to understand whether to reset the
+                # view color!
+                if self.size == subtreeSize:
```
If `self.size` is the full size of the subtree, and is in fact read only, can you encode this info better in the name / code?

lohathe committed:
Apply PEP 8.

lohathe committed:
Minor refactoring for readability.

lohathe committed:
Use Qt to force GUI to process events while heavy loaded.

lohathe committed:
Fix filter matching logic.

No need to specify the selected dir as part of the filters:
matching starts from the children of the user-selected root folder.

lohathe committed:
Fix output management.

(o) notify that filters are not applyed when changing
	or refreshing the root folder
(o) do not clear the list of filters when changing
	the root folder
(o) using QSettings to persist user input

lohathe committed:
Fix uncatched errors for bad regex.

lohathe committed:
Add sorting to tree view.

lohathe committed:
Enhance tree view.

(o) Add per-column sorting
(o) Add context-menu to directly exclude items
